### PR TITLE
Add autodiff library node usable in DacePrograms

### PR DESCRIPTION
**Pull request by orausch**
This adds support for the ``torch.autograd.backward`` function, as well
as access to `.grad` of an array.

Calls to ``torch.autograd.backward`` inserts a new ``BackwardPass``
library node.

Calls to ``.grad`` allocate gradient buffers and write to them via these
library nodes.

Example:
```python
dace_module = DaceModule(dace_module, auto_optimize=False)

@dace
def train_step(x):
    output = dace_module(x)
    loss = np.add.reduce(output, axis=[0, 1])
    loss.backward()
    return loss
```
Outputs:
Screenshot 2022-08-03 at 21 25 54
Which expands to:
Screenshot 2022-08-03 at 21 27 22

### CODE REVIEW

File: daceml/autodiff/backward_pass_generator.py [hunk 1]
```diff
@@ -800,7 +828,7 @@ def _add_gradient_data_descriptor(self, data_name: str):
 
         array = self.sdfg.arrays[data_name]
 
-        if type(array) not in [dt.Scalar, dt.Array, dt.View]:
+        if not isinstance(array, (dt.Scalar, dt.Array, dt.View)):
```
Comment (tbennun) [inline daceml/autodiff/backward_pass_generator.py:831]: this means that References are supported for autodiff as well. This is good, they should be.

File: daceml/autodiff/library/library.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+                for edge in state.in_edges(node):
+                    if edge.data.data == name:
+                        nested_names.add(edge.dst_conn)
+                for edge in state.out_edges(node):
+                    if edge.data.data == name:
+                        nested_names.add(edge.dst_conn)
+
+                for name in nested_names:
+                    nested_desc = node.sdfg.arrays[name]
```
Comment (tbennun) [inline daceml/autodiff/library/library.py:69]: This only goes one level deep. Explain why this is necessary or if we need to traverse further
Also, why do you take the internal descriptor rather than the external one, which is the one that may fit the array better (e.g., grad of A[:, 1])

File: daceml/autodiff/library/python_frontend.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                    state: SDFGState,
+                    self: str,
+                    grad: Optional[str] = None):
+    """
+    Alias for ``torch.autograd.backward(self)``
+    """
+    backward(pv, sdfg, state, self, grad)
+
+
+# FIXME hack to make sure BackwardPass nodes are expanded before other library nodes when using the python frontend
```
Comment (tbennun) [inline daceml/autodiff/library/python_frontend.py:172]: :/

File: daceml/autodiff/library/python_frontend.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    An optimizer that expands the backward pass nodes of an SDFG.
+
+    Since optimizers are called before other transformations (and compilation)
+    in DaceProgram.__call_, this prevents other library nodes from being
+    expanded before the backward pass nodes.
+
+    This is clearly a hack and will be replaced with better way to achieve this.
+    """
+    def optimize(self):
+        expand_nodes(self.sdfg, lambda n: isinstance(n, BackwardPass))
```
Comment (tbennun) [inline daceml/autodiff/library/python_frontend.py:197]: shouldn't it at least call `_current_opt` after this?

File: daceml/torch/module.py [hunk 5]
```diff
@@ -396,21 +397,74 @@ def __sdfg__(self, *args):
...
         inputs = [clean_onnx_name(name) for name in self.dace_model.inputs]
+        grad_buffers = self._add_gradient_buffers()
+        inputs.extend(grad_buffers)
+
         return inputs, []
 
+    @staticmethod
+    def _tensor_from_param(param):
+        t = param.data
+        # for some reason doing .data on a Parameter kills the requires_grad flag
```
Comment (tbennun) [inline daceml/torch/module.py:450]: it's because it is normally used in a no_grad environment to modify the data buffer itself.

**Review by tbennun — COMMENTED** [review]
There are some open questions and one issue but all looks well (apart from the hacks)

File: daceml/autodiff/library/library.py [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+from dace.transformation import transformation as pm
+from dace.transformation.passes import analysis
+from dace.sdfg import graph
+
+from daceml.autodiff import backward_pass_generator as engine, analysis as autodiff_analysis
+from daceml.util.utils import in_edge_with_name
+
+
+@properties.make_properties
+class Array(data.Array):
```
[the remark this answers is not in the document]
Comment (orausch) [reply daceml/autodiff/library/library.py:21]: In dace's `replacements.py`, you can see that the binop replacements match on the operand's `__class__.__name__`. This means that naming this anything else prevents the replacements from matching, and then parameters can't be used for any binops.

File: daceml/autodiff/library/library.py (hunk 2 above)
Comment (orausch) [reply daceml/autodiff/library/library.py:69]: Here's the comment I will add, let me know if this clears it up:
```
 First, check if this array already has a gradient buffer in a nested SDFG
 This happens, for example when pytorch modules are used in the frontend. In that case:
 1. the parser assembles the closure of the module, which adds
    descriptors for all the parameters and their gradients (if they are
    required).
 2. The DaceProgram will then pass these arrays in when the
     DaceProgram is called, using the names from the inner closure.
 3. When parsing the backward nodes, we want the gradient buffers in
    the closure to match the gradient buffers that we pass in. Thus, we
    need to make sure that we use the same name as the NestedSDFG

 Note that we do not currently do any nesting beyond this level,
 because nested modules are converted to one SDFG.
```

File: daceml/autodiff/library/library.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (orausch) [reply daceml/autodiff/library/library.py:21]: > Yes, that I understand, but why is the subclass even necessary as a subclass?

Not sure I understand what you mean, but we need the `ParameterArray` class so that we can mark the arrays that require gradients. Later on this won't be needed if we add a `gradient` field to dace arrays.

File: daceml/autodiff/library/python_frontend.py (hunk 4 above)
Comment (orausch) [reply daceml/autodiff/library/python_frontend.py:197]: Yes I will add that

File: daceml/torch/module.py (hunk 5 above)
Comment (orausch) [reply daceml/torch/module.py:450]: Will update this comment

**Review by tbennun — APPROVED** [review]
